### PR TITLE
ACMP support for value types

### PR DESCRIPTION
**Pull request by tajila**
ACMP compatibility for value types

This adds support for substitutability test in the acmp* bytecodes which
is needed to support value types.

Overview:
-----------

Instead of doing a reference comparison when the ifacmp* bytecodes are
executed with valueType operands, as is the case with reference type
today, valueTypes will perform a structural comparison instead
(substitutability test). This overload of the ifacmp* bytecodes is
required because valueTypes subclass java.lang.Object and use a*
bytecodes. It is possible for the static type of a valueType to be a
reference type (see example below) but the reflexive properties of the
`==` must still be maintained.

```
Object p = (Object) Point.makePoint(1, 2);
if (p == p) //must be true
```

The original acmp (legacy acmp) operation in the ifacmp* bytecodes
performs a simple ref comparison, lhs == rhs. This must be updated to
check if both operands are valueTypes, and  if they are of the same
type. If that is the case a structural comparison must be performed.
Example below:

```
acmpSub(lhs, rhs) {
        if (lhs == rhs) return true
        if ((lhs == NULL) || (rhs == NULL)) return false
        if ((lhs.class == rhs.class) && (lhs.isValue)) return
isSubstitutable(lhs, rhs)
        return false
}
```

The structural comparison must iterate through each field and perform an
equality comparison. If the field is byte, short, char, int the
comparison in icmp_eq is performed. If the field is a long the
comparison in lcmp is performed. If the field is a float an icmp
comparison of `Float.floatToIntBits` of both values is performed. If the
field is a double an lcmp comparison of `Double.doubleToLongBits` of both
values is performed. Note: this behaviour differs from the `float f =
...; f == f` case which performs a fcmp* operation (See Float.equals() in https://docs.oracle.com/javase/9/docs/api/java/lang/Float.html#equals-java.lang.Object-). If the field is a
reference type that is java.lang.Object or an Interface or any type that
may contain a valueType (i.e InlineObject, Object, restricted-abstract
classes or interfaces) or if the field is a value type and both values
are not NULL a recursive acmpSub is performed. For all other types the
legacy acmp is performed. Example is below:

```
isSubstitutable(lhs, rhs) {
        for each primitive perform primitive comparison (icmp, lcmp,
Float.floatToIntBits then icmp, )
        for each ref that is j.l.Object, interface, a valueType and
both operands are not NULL perform acmSub (recursive)
        for all other refs perform legacy acmp
        return result
}
```

There are cases where the JVM may optimize the individual field
comparisons of the substitutability test and simply perform a memcmp
over the range of the operands. This optimization is possible in cases
where the substitutability test will not require recursion and there are
no floating point types.
The acmpSub will check to see if the fast comparison flag exists, if so
it will perform a memcmp with both operands that does something like
memcmp(lhs, rhs, lhs.size).

This is based on Adithya's work in
#8133

Signed-off-by: Tobi Ajila <atobia@ca.ibm.com>

### CODE REVIEW

**Comment by tajila** [comment]
@gacholio please review these changes

**Comment by tajila** [comment]
I did some refactoring, changes are ready now

File: runtime/vm/BytecodeInterpreter.hpp [hunk 1]
```diff
@@ -7323,56 +7324,6 @@ class INTERPRETER_CLASS
...
-	 * Determine if the two objects are substitutable
-	 *
-	 * @param[in] lhs the lhs object of acmp bytecodes
-	 * @param[in] rhs the rhs object of acmp bytecodes
-	 * return true if they are substitutable and false otherwise
-	 */
-	VMINLINE bool
-	acmp(j9object_t lhs, j9object_t rhs)
-	{
-#if defined(J9VM_OPT_VALHALLA_VALUE_TYPES)
```
Comment (gacholio) [inline runtime/vm/BytecodeInterpreter.hpp:7336]: Might be better to assign acmpResult to (lhs == rhs) initially, then test and modify it inside the ifdef so there's only a single return.

File: runtime/vm/BytecodeInterpreter.hpp [hunk 2]
```diff
@@ -48,6 +48,7 @@
 #include "MHInterpreter.hpp"
 #include "ObjectAccessBarrierAPI.hpp"
 #include "ObjectHash.hpp"
+#include "valueTypeHelpers.hpp"
```
Comment (gacholio) [inline runtime/vm/BytecodeInterpreter.hpp:51]: Is there a reason not to captialize here?

File: runtime/vm/BytecodeInterpreter.hpp (hunk 1 above)
Comment (gacholio) [reply runtime/vm/BytecodeInterpreter.hpp:7336]: Once again, I can't tell deletion from addition...

File: runtime/vm/valueTypeHelpers.hpp [hunk 3]
```diff
@@ -37,14 +43,59 @@ class ValueTypeHelpers {
...
 
 public:
-	static bool
-	isSubstitutable(J9VMThread *currentThread, j9object_t lhs, j9object_t rhs, UDATA startOffset, J9Class *clazz)
+	static VMINLINE bool
+	acmp(J9VMThread *currentThread, MM_ObjectAccessBarrierAPI objectAccessBarrier, j9object_t lhs, j9object_t rhs)
 	{
-		return false;
+#if defined(J9VM_OPT_VALHALLA_VALUE_TYPES)
+		bool acmpResult = false;
```
Comment (gacholio) [inline runtime/vm/valueTypeHelpers.hpp:81]: Assign inital value to (rhs == lhs) and test inside the ifdef, resulting in a single return at the end.

File: runtime/vm/valueTypeHelpers.hpp [hunk 4]
```diff
@@ -23,6 +23,12 @@
 #if !defined(VALUETYPEHELPERS_HPP_)
 #define VALUETYPEHELPERS_HPP_
 
+#include "j9.h"
+#include "fltconst.h"
+
+#include "ObjectAccessBarrierAPI.hpp"
+#include "VMHelpers.hpp"
+
 class ValueTypeHelpers {
```
Comment (gacholio) [inline runtime/vm/valueTypeHelpers.hpp:32]: To be consistent, this should be VM_ValueTypeHelpers.

**Comment by gacholio** [comment]
Jenkins test sanity,extended xlinuxval jdknext

**Pull request merged**